### ISSUE TITLE
[ERROR:flutter/lib/ui/ui_dart_state.cc(166)] Unhandled Exception: setState() called after dispose(): _StreamBuilderBaseState

### ISSUE THREAD

**Issue by angel1st** [simolus3/moor#766]
Hey @simolus3, I bumped into [the following issue](https://stackoverflow.com/questions/63429252/flutter-unhandled-exception-setstate-called-after-dispose-and-streambuil) and since I don't have `setState` in my code and the only workable snipped is invocation of the Moor database stream, I wonder if this might somehow be related to your package (no offense :-))
Thanks in advance for your time!

**Comment by simolus3** [comment]
Thanks for reporting this. The `moor` package does not depend on Flutter, so I can assure it doesn't call `setState` itself :)

Async stack traces in Dart can be a bit confusing at times, but I think the trace you posted means that, somewhere, you're listening to a moor stream and call `setState` in the listener. Here, that's indeed the `StreamBuilder`. I assume that `StreamBuilder` properly cancels its subscription, so this might still be a moor bug.
I recently fixed a stream bug, so you can you try to use moor version `3.3.1`? That probably doesn't fix the problem, but having a stack trace created with that version would help.

**Comment by angel1st** [comment]
Thanks, man. I will upgrade to `3.3.1` and let you know in any case.

**Comment by angel1st** [comment]
Here is the stack trace with `3.3.1`:

> E/flutter (  794): [ERROR:flutter/lib/ui/ui_dart_state.cc(166)] Unhandled Exception: setState() called after dispose(): _StreamBuilderBaseState<List<ScheduledActivity>, AsyncSnapshot<List<ScheduledActivity>>>#2cc54(lifecycle state: defunct, not mounted)
E/flutter (  794): This error happens if you call setState() on a State object for a widget that no longer appears in the widget tree (e.g., whose parent widget no longer includes the widget in its build). This error can occur when code calls setState() from a timer or an animation callback.
E/flutter (  794): The preferred solution is to cancel the timer or stop listening to the animation in the dispose() callback. Another solution is to check the "mounted" property of this object before calling setState() to ensure the object is still in the tree.
E/flutter (  794): This error might indicate a memory leak if setState() is being called because another object is retaining a reference to this State object after it has been removed from the tree. To avoid memory leaks, consider breaking the reference to this object during dispose().
E/flutter (  794): #0      State.setState.<anonymous closure> (package:flutter/src/widgets/framework.dart:1204:9)
E/flutter (  794): #1      State.setState (package:flutter/src/widgets/framework.dart:1239:6)
E/flutter (  794): #2      _StreamBuilderBaseState._subscribe.<anonymous closure> (package:flutter/src/widgets/async.dart:140:9)
E/flutter (  794): #3      new _StartWithValueSubscription.<anonymous closure> (package:moor/src/utils/start_with_value_transformer.dart:58:20)
E/flutter (  794): #4      _rootRun (dart:async/zone.dart:1182:47)
E/flutter (  794): #5      _CustomZone.run (dart:async/zone.dart:1093:19)
E/flutter (  794): #6      _CustomZone.runGuarded (dart:async/zone.dart:997:7)
E/flutter (  794): #7      _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1037:23)
E/flutter (  794): #8      _rootRun (dart:async/zone.dart:1190:13)
E/flutter (  794): #9      _CustomZone.run (dart:async/zone.dart:1093:19)
E/flutter (  794): #10     _CustomZone.runGuarded (dart:async/zone.dart:997:7)
E/flutter (  794): #11     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1037:23)
E/flutter (  794): #12     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
E/flutter (  794): #13     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
E/flutter (  794): 

This is my `flutter doctor -v` output:

>[√] Flutter (Channel stable, 1.20.2, on Microsoft Windows [Version 10.0.18362.1016], locale en-US)
    • Flutter version 1.20.2 at d:\DevTools\flutter
    • Framework revision bbfbf1770c (2 days ago), 2020-08-13 08:33:09 -0700
    • Engine revision 9d5b21729f
    • Dart version 2.9.1
[√] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at D:\DevTools\Android\Sdk
    • Platform android-29, build-tools 29.0.3
    • ANDROID_SDK_ROOT = D:\DevTools\Android\Sdk
    • Java binary at: D:\DevTools\android-studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.
[√] Android Studio (version 4.0)
    • Android Studio at D:\DevTools\android-studio
    • Flutter plugin version 47.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
[√] Connected device (1 available)
    • Android SDK built for x86 (mobile) • emulator-5554 • android-x86 • Android 10 (API 29) (emulator)
• No issues found!

**Comment by angel1st** [comment]
> Async stack traces in Dart can be a bit confusing at times, but I think the trace you posted means that, somewhere, you're listening to a moor stream and call setState in the listener. Here, that's indeed the StreamBuilder

At the moment I remove that particular `StreamBuilder`:

```
StreamBuilder<List<ScheduledActivity>>(
      stream: RepositoryProvider.of<AppDatabase>(context)
          .scheduledActivitiesDao
          .watchActivitiesByPlant(widget.plantId),
      builder: (context, snapshot) {})
```
as explained in my original post, everything seems just fine. Hence I claim the issue is somehow related to the StreamBuilder, however, I cannot figure out how.
Anyway - thanks for the feedback!

**Comment by davidmartos96** [comment]
There is one issue I see with your code. You are creating a new Stream of data every time the build method is called. Streams and Futures provided to StreamBuilders and FutureBuilders should be created once on initState and update accordingly in the didUpdateWidget/didchangedependencies methods.

Even though that is a "bug" I'm not sure it's related to the error you are having. I would try to fix that first though.

**Comment by angel1st** [comment]
To complete the case - this is `watchActivitiesByPlant` method:

```
Stream<List<ScheduledActivity>> watchActivitiesByPlant(int plantId) =>
      (select(scheduledActivities)..where((sa) => sa.plantId.equals(plantId)))
          .watch();
```

and `ScheduledActivities` table:

```
class TimeConverter extends TypeConverter<mat.TimeOfDay, DateTime> {
  const TimeConverter();
  @override
  mat.TimeOfDay mapToDart(DateTime date) {
    if (date == null) {
      return null;
    }
    return mat.TimeOfDay.fromDateTime(date);
  }
  @override
  DateTime mapToSql(mat.TimeOfDay time) {
    if (time == null) {
      return null;
    }
    return DateTime(1969, 1, 1, time.hour, time.minute);
  }
}

@DataClassName("ScheduledActivity")
class ScheduledActivities extends Table {
  IntColumn get id => integer().autoIncrement()();
  IntColumn get plantId =>
      integer().customConstraint('REFERENCES plants(id) ON DELETE CASCADE')();
  IntColumn get activityType =>
      intEnum<ActivityTypesEnum>().withDefault(Constant(0))();
  DateTimeColumn get lastExecuted =>
      dateTime().withDefault(currentDateAndTime)();
  DateTimeColumn get dueDate =>
      dateTime().withDefault(currentDateAndTime)();
  IntColumn get count => integer().withDefault(Constant(1))();
  IntColumn get periodType => intEnum<PeriodsEnum>().withDefault(Constant(0))();
  DateTimeColumn get timeReminder => dateTime().map(const TimeConverter()).nullable()();
}

```
as you can see it's very straightforward...

**Comment by angel1st** [comment]
@davidmartos96 - thanks for the hand. That was it!

The steps:

1. Add `Stream<List<ScheduledActivity>> _stream;`
2. Set its value here:
```
  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    _stream = RepositoryProvider.of<AppDatabase>(context)
        .scheduledActivitiesDao
        .watchActivitiesByPlant(widget.plantId);
  }

```
3. Replace `RepositoryProvider.of<AppDatabase>(context).scheduledActivitiesDao.watchActivitiesByPlant(widget.plantId),` with `stream: _stream,`

Thanks again @davidmartos96 - you saved my day!

**Comment by davidmartos96** [comment]
@angel1st Glad you found it! It's a common pitfall with StreamBuilders.
Double check if `didChangeDependencies` is the right place for the Stream creation. 
Your stream depends on `widget.plantId`, which when changes, `didUpdateWidget` will be triggered.
So you may want to do something like this.

```dart
void createStream() {
  _stream = RepositoryProvider.of<AppDatabase>(context)
        .scheduledActivitiesDao
        .watchActivitiesByPlant(widget.plantId);
}

void initState() {
  createStream();
}

void didUpdateWidget(MyWidget oldWidget) {
  if (oldWidget.plantId != widget.plantId) {
    createStream();  
  }
}
```

**Comment by simolus3** [comment]
FWIW, I think this might have been another moor bug with streams. When calling `cancel` or `pause` on a stream subscription by moor, it might emit cached data after that. I've fixed this on `develop`.

> It's a common pitfall with StreamBuilders

Out of interest, do you have a source explaining why we can't create the stream in the `build` method (apart from it being potentially less efficient)? I don't think it should change the behavior if the stream isn't buggy (like here).

**Comment by davidmartos96** [comment]
@simolus3 That's why I said earlier that I wasn't sure it was the cause for the error. It shouldn't be, but I see you found some issue witht the streams in the moor code.
I said we can't create streams/futures in the build method because it is a side effect, and even though side effects should not crash the app, may cause unwanted behavior when rotating the screen, parent widget updates, theme changes etc...
I may have to reword it to "bad practice" instead :smile: 
Here is a nice post on Stack Overflow by Remi (creator of multiple Flutter packages) about it.

https://stackoverflow.com/a/52249579/5617722

**Comment by angel1st** [comment]
@davidmartos96 - thanks for the note, man! In this particular case, however, I believe `didChangeDependencies` is the right place.
If we have a look at the [didChangeDependencies](https://api.flutter.dev/flutter/widgets/SingleTickerProviderStateMixin/didChangeDependencies.html), you can read the following:
> For example, if the previous call to build referenced an InheritedWidget that later changed, the framework would call this method to notify this object about the change.

and, since I am using [bloclibrary](https://bloclibrary.dev/#/) once the BLoC state updates, it will invoke `didChangeDependencies`.

Thanks anyway!

BTW - looking at your code above, please be aware this code snippet:

```
void initState() {
  createStream();
}

```
has to be changed to

```
void initState() {
  if (mounted) createStream();
} 

```
since, `createStream()` relies on `contex`.

P.S. The link you shared is very informative, thanks for it!
